### PR TITLE
[dmd-cxx] Add root/system.h header for wrapping system includes

### PR DESCRIPTION
**Pull request by ibuclaw**
Portability of the front-end is not well covered for older targets.

For instance, while Solaris 11 can build just fine, Solaris 10 fails with:
```
/vol/gcc/src/hg/trunk/local/gcc/d/dmd/arrayop.c: In member function 'virtual void buildArrayIdent(Expression*, OutBuffer*, Expressions*)::BuildArrayIdentVisitor::visit(BinAssignExp*)':
/vol/gcc/src/hg/trunk/local/gcc/d/dmd/arrayop.c:336:29: error: 's' may be used uninitialized in this function [-Werror=maybe-uninitialized]
  336 |             buf->writestring(s);
      |             ~~~~~~~~~~~~~~~~^~~

/vol/gcc/src/hg/trunk/local/gcc/d/dmd/ctfeexpr.c: In function 'int comparePointers(TOK, Expression*, dinteger_t, Expression*, dinteger_t)':
/vol/gcc/src/hg/trunk/local/gcc/d/dmd/ctfeexpr.c:925:13: error: 'n' may be used uninitialized in this function [-Werror=maybe-uninitialized]
  925 |         int n;
      |     
```

This is because assert is not marked as noreturn in `<assert.h>`.

GCC itself provides a couple of headers that wrap around vagaries of the host system to provide a consistent interface for maximum portability.

This patch then moves most front-end system includes into a central place, allowing gdc to override and provide its own copy of this header that includes the gcc headers, or provide gcc-specific portability fixes to the C++ sources.

I need to first check this against gdc to make sure it actually fits well.

### CODE REVIEW

**Comment by ibuclaw** [comment]
All malloc and friends are all marked as `#pragma GCC poison`, so have added ifdefs to forward to the correct xmalloc and friends if compiling gdc.

**Review by jacob-carlborg — APPROVED** [review]

File: src/objc_glue.c [hunk 1]
```diff
@@ -145,7 +145,9 @@ Symbol *objc_getMethVarName(const char *s, size_t len)
 
 Symbol *objc_getMethVarName(Identifier *ident)
 {
-    return objc_getMethVarName(ident->string, ident->len);
+    const char *id = ident->toChars();
+    size_t id_length = strlen(id);
+    return objc_getMethVarName(id, id_length);
```
Comment (jacob-carlborg) [inline src/objc_glue.c]: Why is this change necessary?

Comment (ibuclaw) [reply src/objc_glue.c]: https://auto-tester.puremagic.com/show-run.ghtml?projectid=17&runid=3396619&isPull=true

Whilst waiting, just curious if OSX builds would pass if the compiler error were fixed.

File: src/root/dsystem.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * Distributed under the Boost Software License, Version 1.0.
+ * http://www.boost.org/LICENSE_1_0.txt
+ * https://github.com/dlang/dmd/blob/master/src/dmd/root/dsystem.h
+ */
+
+#pragma once
+
+// Get common system includes from the host.
+
+#define POSIX (__linux__ || __GLIBC__ || __gnu_hurd__ || __APPLE__ || __FreeBSD__ || __OpenBSD__ || __sun)
```
Comment (jacob-carlborg) [inline src/root/dsystem.h]: If this is supposed to be up to date with the master branch then there's DragonFly BSD as well.

File: src/objc_glue.c (hunk 1 above)
Comment (jacob-carlborg) [reply src/objc_glue.c]: I see now that this might be necessary since `Identifier` is now holding a D string.

File: src/root/dsystem.h (hunk 2 above)
Comment (ibuclaw) [reply src/root/dsystem.h]: I don't see DragonflyBSD in master.

Comment (jacob-carlborg) [reply src/root/dsystem.h]: Seems like you removed it recently: https://github.com/dlang/dmd/commit/d1bfdbca.

File: src/objc_glue.c (hunk 1 above)
Comment (jacob-carlborg) [reply src/objc_glue.c]: The fields in `Identifier` used to public at one point in time.

File: src/root/dsystem.h (hunk 2 above)
Comment (ibuclaw) [reply src/root/dsystem.h]: Ah, it's in other sources.  Probably worth putting it in a separate after this if still needed.

Comment (ibuclaw) [reply src/root/dsystem.h]: Oh, heh, copy-pasto.

**Comment by jacob-carlborg** [comment]
> Everything else ok?

Looks good, nothing else that I can see.

**Review by jacob-carlborg — APPROVED** [review]
